File: java/src/main/BinaryTreePreOrderTraversal.java
package main;

import main.util.TreeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 Given a binary tree, return the preorder traversal of its nodes' values.

 For example:
 Given binary tree {1,#,2,3},
 1
 \
 2
 /
 3
 return [1,2,3].

 Note: Recursive solution is trivial, could you do it iteratively?
 */
public class BinaryTreePreOrderTraversal {
    //This is a DFS
    List<Integer> result = new ArrayList<Integer>();
    public List<Integer> preorderTraversal(TreeNode root) {
        if(root == null)
            return result;

        Stack<TreeNode> stk = new Stack<TreeNode>();
        stk.push(root);
        while(!stk.empty()) {
            TreeNode tmp = stk.pop();
            result.add(tmp.val);

            if(tmp.right != null) {
                stk.push(tmp.right);
            }

            if(tmp.left != null) {
                stk.push(tmp.left);
            }
        }
        return result;
    }
}

File: python/binaryTreePreorderTraversal.py
__author__ = 'don'
"""
Given a binary tree, return the preorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [1,2,3].

Note: Recursive solution is trivial, could you do it iteratively?
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {integer[]}
    def preorderTraversal(self, root):
        r = []
        self.helper(root, r)
        return r

    def helper(self, root, r):
        if root is None:
            return
        r.append(root.val)
        self.helper(root.left, r)
        self.helper(root.right, r)

